refactor: asset-grid, actions

midzelis opened the pull request:
Improve general timeline components, breaking up the huge components into smaller chunks. Will try to group changes into smaller commits within this PR.

midzelis committed:
Extract asset viewer navigation logic into separate components

• Move navigation handlers (next, previous, random, close) from asset-grid to dedicated asset-viewer-actions component
• Extract asset action handlers (archive, stack, favorite) into reusable module  
• Create asset-viewer-and-actions wrapper to compose viewer with navigation actions

midzelis committed:
Extract asset selection logic into asset-date-group-actions component

• Move multi-asset selection handlers (shift-click range selection) from asset-grid
• Extract date group selection logic for bulk selecting assets by date
• Consolidate keyboard event handling for shift key detection
• Clean up asset-grid by removing ~200 lines of selection-related code

midzelis committed:
Extract keyboard shortcuts and asset actions into asset-grid-actions component

• Move all keyboard shortcut handling and focus navigation logic from asset-grid
• Extract asset operation handlers (delete, force delete, stack, archive)
• Move delete confirmation dialog and date picker to new component
• Consolidate ~230 lines of action-related code into dedicated module

jrasm91 commented:
I'd probably try to do the component refactoring first and move it later, otherwise you will constantly end up with conflicts for anything and everything that's also changed in the asset-grid file.

midzelis committed:
Extract asset grid core logic into asset-grid-without-scrubber component

• Move all timeline rendering and scrolling logic to new component
• Keep scrubber logic and interaction handlers in original asset-grid 
• Use composition pattern with header snippet for scrubber integration
• Simplify asset-grid to ~450 lines by extracting ~390 lines of core logic

midzelis committed:
Fix and enhance navigate to time functionality

• Add i18n support for navigate button and dialog title
• Update ChangeDate component to return DateTime object alongside ISO string
• Fix closest date finding algorithm to handle cases where exact month doesn't exist
• Add navigate to time (g) shortcut to keyboard shortcuts modal

midzelis committed:
Refactor date group actions from Svelte component to TypeScript class

• Convert asset-date-group-actions.svelte to date-group-actions-lib.svelte.ts class
• Remove complex prop binding between asset-grid-without-scrubber and asset-date-group
• Use class instance in asset-date-group with direct method access

midzelis committed:
Rename scroll compensation functions for clarity

• Rename handleScrollCompensation to scrollCompensation
• Rename scrollTop prop to onScrollToTop to match event handler pattern
• Add onScrollCompensation alias for consistency with other event handlers

midzelis committed:
Split asset date group into (temp) generic and selection-aware components

• Rename asset-date-group.svelte to asset-date-group-comp.svelte as generic base
• Create asset-date-group-selection-aware.svelte wrapper for selection logic
• Extract selection-specific callbacks into separate handlers
• Update onDateGroupSelect to onDayGroupSelect with DayGroup parameter

midzelis committed:
Inline DateGroupActionLib logic into asset-date-group-selection-aware (temp)

• Remove DateGroupActionLib dependency and inline all selection logic
• Implement proper selection state predicates instead of placeholder true values
• Move keyboard event handlers and selection logic directly into component

midzelis committed:
Replace AssetDateGroup with AssetDateGroupSelectionAware throughout - merge/replace temp

• Update asset-grid-without-scrubber to use AssetDateGroupSelectionAware
• Remove original asset-date-group.svelte component
• Remove date-group-actions-lib.svelte.ts class

midzelis committed:
Merge asset viewer components into asset-grid-asset-viewer

• Rename asset-viewer-actions.svelte to asset-grid-asset-viewer.svelte
• Move AssetViewer rendering from asset-viewer-and-actions into asset-grid-asset-viewer
• Remove asset-viewer-and-actions.svelte intermediate wrapper
• Remove unused isShowDeleteConfirmation prop

midzelis committed:
Rename and relocate asset-grid-without-scrubber to timeline-viewer

• Rename asset-grid-without-scrubber.svelte to timeline-viewer.svelte
• Move timeline-viewer.svelte to lib/components/timeline-viewer/
• Move skeleton.svelte to lib/components/timeline-viewer/
• Update imports in asset-grid.svelte

midzelis committed:
Rename and relocate timeline-related components

• Move asset-grid-asset-viewer to timeline-asset-viewer in timeline-viewer/
• Move asset-date-group-selection-aware to selectable-timeline-day in timeline-viewer/timeline-day/
• Move asset-date-group-comp to timeline-day in timeline-viewer/timeline-day/
• Update all imports in timeline-viewer.svelte

midzelis committed:
Rename asset-grid to base-timeline and reorganize timeline components

• Rename asset-grid.svelte to base-timeline.svelte
• Move asset-grid-actions.svelte to timeline-keyboard-actions.svelte in timeline-viewer/actions/
• Move delete-asset-dialog.svelte to timeline-viewer/actions/
• Update all imports across routes and components

midzelis committed:
Reorganize timeline components into final directory structure

• Move timeline-viewer to timeline/base-components/base-timeline-viewer
• Move base-timeline to timeline/base-components/base-timeline
• Move skeleton to timeline/base-components/
• Move timeline-day to timeline/base-components/
• Move selectable-timeline-day to timeline/internal-components/
• Move timeline-asset-viewer to timeline/internal-components/
• Move timeline actions to timeline/actions/
• Update all imports to reference new locations

midzelis committed:
Create Timeline facade component to unify timeline usage

• Create timeline/timeline.svelte as main entry point for timeline functionality
• Combine BaseTimeline, TimelineKeyboardActions, and TimelineAssetViewer
• Update all route imports from base-timeline to use Timeline component
• Move scrubber.svelte to timeline/base-components/
• Fix timeline-keyboard-actions date handling from result.dateTime to result.date
• Clean up unused imports and props

midzelis committed:
Fix code quality issues in timeline components

• Fix variable naming consistency (leadout → leadOut)
• Remove unused props from timeline-asset-viewer interface
• Add styleMarginRightOverride prop for dynamic margin control
• Simplify skeleton component styling

midzelis committed:
Extract HMR logic into reusable component

• Create dedicated Hmr component for hot module replacement handling
• Remove inline HMR code from base-timeline-viewer
• Simplify timeline viewer by delegating HMR concerns to specialized component

danieldietzler commented on web/src/lib/components/timeline/internal-components/selectable-timeline-day.svelte:
```diff
@@ -0,0 +1,262 @@
+        }
+      }
+
+      // Update date group selection in range [start,end]
+      started = false;
+      for (const monthGroup of timelineManager.months) {
```
And now you're searching that same bucket again? 👀 
Definitely even more of a reason to do my suggestion above.

danieldietzler commented on web/src/lib/components/timeline/base-components/base-timeline-viewer.svelte:
```diff
@@ -0,0 +1,297 @@
+    showArchiveIcon = false,
+    styleMarginRightOverride,
+    isShowDeleteConfirmation = $bindable(false),
+    onSelect = (_: TimelineAsset) => void 0,
```
Oh yeah sorry I'm lost I meant `() => {}`

midzelis committed:
Improve consistency between scrollTop and scrollTo

midzelis committed:
Move updateSlidingWindow into onScroll

midzelis committed:
refactor hmr a bit more

midzelis committed:
refactor scrollCompensation a bit more

midzelis committed:
refactor/split handleTimelineScroll into smaller functions

midzelis committed:
Minor scrubber refactor

midzelis committed:
refactor - improve timeline readability, general refactoring, renaming variables, functions, props, etc

midzelis committed:
rename timeline-day to timeline-month

alextran1502 commented:
Hi @midzelis, thank you for working on this! Do you think it is also a good time to add tests? As this is a pretty critical part of the application, going through a big refactor without tests is also very scary :D

midzelis commented:
> Hi @midzelis, thank you for working on this! Do you think it is also a good time to add tests? As this is a pretty critical part of the application, going through a big refactor without tests is also very scary :D

Yes, tests are absolutely needed. In this refactor, this refactor almost exclusively touches the UI/browser svelte components. The timeline components are notoriously difficult to unit test - due to poor mocking/simulation of DOM events, especially around scrolling, mouse dragging, and the like. For ref, here's what svelte docs say about component testing: https://svelte.dev/docs/svelte/testing#Unit-and-integration-testing-using-Vitest-Component-testing

What would make more sense for testing these refactored components - is to test these in an "end-to-end" style. That also comes with difficulties, as there are no existing test suites that set up a large number of assets, that span multiple months and years, that would be a prerequisite to minimally cover the components using playwright (or similar). 

Note - an earlier refactor in this area (AssetManager -> TimelineManager) already refactored most of the non-browser specific UI "logic". This can and should be unit tested (but this logic later was not touched by this refactor). 

So I'm in agreement with you - though I think there would need to be dedicated PRs to establish a multi-month series of assets that can be then used by the timeline -- before those tests can be written. I feel its likely too many changes to include in this PR. The timeline has (luckily) survived two significant rewrites already and this refactor is much smaller compared to those.

midzelis committed:
Review comments

midzelis commented on web/src/lib/components/timeline/base-components/hmr.svelte:
```diff
@@ -0,0 +1,16 @@
+
+  onMount(() => {
+    if (import.meta && import.meta?.hot) {
+      import.meta.hot?.on('vite:afterUpdate', onAfterUpdate);
```
Oh, I see why - if I don't do this, this is actually NOT an eslint error, rather, it comes from tsc: 
```
Object is possibly 'undefined'.ts(2532)
```
-- I'll add an additional check, just in case.

danieldietzler replied:
I don't understand. You literally check `if (import.meta?.hot)`, how can that be undefined then?

midzelis committed:
readability improvements in scroll function

midzelis committed:
use binary search for perf, refactor, improve readability

midzelis committed:
Inline, combine functions, lint
